File: server/internal/ws/events.go
package ws

import "encoding/json"

// Client -> Server events
const (
	EventMessageSend    = "message:send"
	EventMessageEdit    = "message:edit"
	EventMessageDelete  = "message:delete"
	EventTypingStart    = "typing:start"
	EventTypingStop     = "typing:stop"
	EventPresenceUpdate = "presence:update"
	EventChannelJoin    = "channel:join"
	EventChannelLeave   = "channel:leave"

	// Thread client -> server events
	EventThreadMessageSend   = "thread:message:send"
	EventThreadMessageEdit   = "thread:message:edit"
	EventThreadMessageRemove = "thread:message:remove"

	// Voice client -> server events
	EventVoiceJoin         = "voice:join"
	EventVoiceLeave        = "voice:leave"
	EventVoiceAnswer       = "voice:answer"
	EventVoiceICECandidate = "voice:ice-candidate"
	EventVoiceMute         = "voice:mute"
	EventVoiceDeafen       = "voice:deafen"
	EventVoiceScreenStart  = "voice:screen-start"
	EventVoiceScreenStop   = "voice:screen-stop"
	EventVoiceWebcamStart  = "voice:webcam-start"
	EventVoiceWebcamStop   = "voice:webcam-stop"

	// Voice E2E key distribution (client <-> server relay)
	EventVoiceKeyDistribute = "voice:key-distribute"
)

// Server -> Client events
const (
	EventMessageNew      = "message:new"
	EventMessageUpdated  = "message:updated"
	EventMessageDeleted  = "message:deleted"
	EventTypingIndicator = "typing:indicator"
	EventPresenceChanged = "presence:changed"
	EventMemberJoined    = "member:joined"
	EventMemberLeft      = "member:left"
	EventChannelCreated  = "channel:created"
	EventChannelUpdated  = "channel:updated"
	EventChannelDeleted  = "channel:deleted"
	EventError           = "error"

	// Thread events
	EventThreadCreated        = "thread:created"
	EventThreadMessageNew     = "thread:message:new"
	EventThreadMessageUpdated = "thread:message:updated"
	EventThreadMessageDeleted = "thread:message:deleted"
	EventThreadUpdated        = "thread:updated"

	// Reaction events
	EventReactionAdded   = "reaction:added"
	EventReactionRemoved = "reaction:removed"

	// Voice server -> client events
	EventVoiceOffer      = "voice:offer"
	EventVoiceICEOut     = "voice:ice-candidate"
	EventVoiceUserJoined = "voice:user-joined"
	EventVoiceUserLeft   = "voice:user-left"
	EventVoiceSpeaking   = "voice:speaking"
	EventVoiceState      = "voice:state"
	EventVoiceMuteUpdate    = "voice:mute-update"
	EventVoiceScreenUpdate  = "voice:screen-update"
	EventVoiceWebcamUpdate  = "voice:webcam-update"

	// DM events
	EventDMMessageNew     = "dm:message:new"
	EventDMMessageUpdated = "dm:message:updated"
	EventDMMessageDeleted = "dm:message:deleted"
	EventDMTypingStart    = "dm:typing:start"
	EventDMTypingStop     = "dm:typing:stop"
	EventDMCreated        = "dm:created"

	// Request/Response events (client requests data over WS)
	EventRequest  = "request"
	EventResponse = "response"

	// Ping/Pong for latency measurement
	EventPing = "ping"
	EventPong = "pong"
)

// Event is the wire format for all WebSocket messages.
type Event struct {
	Type    string          `json:"type"`
	Payload json.RawMessage `json:"payload"`
}

// MakeEvent creates an Event with a JSON-encoded payload.
func MakeEvent(eventType string, payload interface{}) (Event, error) {
	data, err := json.Marshal(payload)
	if err != nil {
		return Event{}, err
	}
	return Event{Type: eventType, Payload: data}, nil
}

type MessageSendPayload struct {
	ChannelID string `json:"channel_id"`
	Content   string `json:"content"`
	Type      string `json:"type"`
	ThreadID  string `json:"thread_id"`
}

type MessageEditPayload struct {
	MessageID string `json:"message_id"`
	ChannelID string `json:"channel_id"`
	Content   string `json:"content"`
}

type MessageDeletePayload struct {
	MessageID string `json:"message_id"`
	ChannelID string `json:"channel_id"`
}

type ThreadMessageSendPayload struct {
	ThreadID string `json:"thread_id"`
	Content  string `json:"content"`
	Nonce    string `json:"nonce"`
}

type ThreadMessageEditPayload struct {
	ThreadID  string `json:"thread_id"`
	MessageID string `json:"message_id"`
	Content   string `json:"content"`
}

type ThreadMessageRemovePayload struct {
	ThreadID  string `json:"thread_id"`
	MessageID string `json:"message_id"`
}

type TypingPayload struct {
	ChannelID string `json:"channel_id"`
	UserID    string `json:"user_id"`
	Username  string `json:"username"`
}

type PresenceUpdatePayload struct {
	UserID     string `json:"user_id"`
	StatusType string `json:"status_type"`
	StatusText string `json:"status_text"`
}

type MessageNewPayload struct {
	ID        string `json:"id"`
	ChannelID string `json:"channel_id"`
	AuthorID  string `json:"author_id"`
	Username  string `json:"username"`
	Content   string `json:"content"`
	Type      string `json:"type"`
	ThreadID  string `json:"thread_id"`
	CreatedAt string `json:"created_at"`
}

type ErrorPayload struct {
	Message string `json:"message"`
}

// DM event payloads

type DMMessageNewPayload struct {
	ID          string `json:"id"`
	DMChannelID string `json:"dm_channel_id"`
	AuthorID    string `json:"author_id"`
	Username    string `json:"username"`
	Content     string `json:"content"`
	Type        string `json:"type"`
	CreatedAt   string `json:"created_at"`
}

type DMMessageUpdatedPayload struct {
	ID          string `json:"id"`
	DMChannelID string `json:"dm_channel_id"`
	Content     string `json:"content"`
	EditedAt    string `json:"edited_at"`
}

type DMMessageDeletedPayload struct {
	ID          string `json:"id"`
	DMChannelID string `json:"dm_channel_id"`
}

type DMTypingPayload struct {
	DMChannelID string `json:"dm_channel_id"`
	UserID      string `json:"user_id"`
	Username    string `json:"username"`
}

type DMCreatedPayload struct {
	ID        string   `json:"id"`
	TeamID    string   `json:"team_id"`
	Type      string   `json:"type"`
	MemberIDs []string `json:"member_ids"`
	CreatedAt string   `json:"created_at"`
}

// Thread event payloads

type ThreadCreatedPayload struct {
	ID              string  `json:"id"`
	ChannelID       string  `json:"channel_id"`
	ParentMessageID string  `json:"parent_message_id"`
	TeamID          string  `json:"team_id"`
	CreatorID       string  `json:"creator_id"`
	Title           string  `json:"title"`
	CreatedAt       string  `json:"created_at"`
}

type ThreadMessageNewPayload struct {
	ID        string `json:"id"`
	ThreadID  string `json:"thread_id"`
	ChannelID string `json:"channel_id"`
	AuthorID  string `json:"author_id"`
	Content   string `json:"content"`
	Type      string `json:"type"`
	CreatedAt string `json:"created_at"`
}

type ThreadMessageUpdatedPayload struct {
	ID       string `json:"id"`
	ThreadID string `json:"thread_id"`
	Content  string `json:"content"`
	EditedAt string `json:"edited_at"`
}

type ThreadMessageDeletedPayload struct {
	ID       string `json:"id"`
	ThreadID string `json:"thread_id"`
}

type ThreadUpdatedPayload struct {
	ID           string  `json:"id"`
	Title        string  `json:"title"`
	MessageCount int     `json:"message_count"`
	LastMessageAt *string `json:"last_message_at"`
}

// Reaction event payload
type ReactionPayload struct {
	MessageID string `json:"message_id"`
	UserID    string `json:"user_id"`
	Emoji     string `json:"emoji"`
	ChannelID string `json:"channel_id"`
}

// Voice event payloads

type VoiceJoinPayload struct {
	ChannelID string `json:"channel_id"`
}

type VoiceLeavePayload struct {
	ChannelID string `json:"channel_id"`
}

type VoiceAnswerPayload struct {
	ChannelID string `json:"channel_id"`
	SDP       string `json:"sdp"`
}

type VoiceICECandidatePayload struct {
	ChannelID string `json:"channel_id"`
	Candidate string `json:"candidate"`
	SDPMid    string `json:"sdp_mid"`
	SDPMLine  uint16 `json:"sdp_mline_index"`
}

type VoiceMutePayload struct {
	ChannelID string `json:"channel_id"`
	Muted     bool   `json:"muted"`
}

type VoiceDeafenPayload struct {
	ChannelID string `json:"channel_id"`
	Deafened  bool   `json:"deafened"`
}

type VoiceOfferPayload struct {
	ChannelID string `json:"channel_id"`
	SDP       string `json:"sdp"`
}

type VoiceUserJoinedPayload struct {
	ChannelID     string `json:"channel_id"`
	UserID        string `json:"user_id"`
	Username      string `json:"username"`
	Muted         bool   `json:"muted"`
	Deafened      bool   `json:"deafened"`
	ScreenSharing bool   `json:"screen_sharing"`
	WebcamSharing bool   `json:"webcam_sharing"`
}

type VoiceUserLeftPayload struct {
	ChannelID string `json:"channel_id"`
	UserID    string `json:"user_id"`
}

type VoiceSpeakingPayload struct {
	ChannelID string `json:"channel_id"`
	UserID    string `json:"user_id"`
	Speaking  bool   `json:"speaking"`
}

type VoiceStatePayload struct {
	ChannelID string      `json:"channel_id"`
	Peers     interface{} `json:"peers"`
}

type VoiceMuteUpdatePayload struct {
	ChannelID string `json:"channel_id"`
	UserID    string `json:"user_id"`
	Muted     bool   `json:"muted"`
	Deafened  bool   `json:"deafened"`
}

type VoiceScreenStartPayload struct {
	ChannelID string `json:"channel_id"`
}

type VoiceScreenStopPayload struct {
	ChannelID string `json:"channel_id"`
}

type VoiceScreenUpdatePayload struct {
	ChannelID string `json:"channel_id"`
	UserID    string `json:"user_id"`
	Sharing   bool   `json:"sharing"`
}

type VoiceWebcamStartPayload struct {
	ChannelID string `json:"channel_id"`
}

type VoiceWebcamStopPayload struct {
	ChannelID string `json:"channel_id"`
}

type VoiceWebcamUpdatePayload struct {
	ChannelID string `json:"channel_id"`
	UserID    string `json:"user_id"`
	Sharing   bool   `json:"sharing"`
}

// VoiceKeyDistributePayload relays encrypted voice keys to channel participants.
// The server cannot read the key contents — they are encrypted per-recipient.
type VoiceKeyDistributePayload struct {
	ChannelID     string            `json:"channel_id"`
	SenderID      string            `json:"sender_id"`
	KeyID         int               `json:"key_id"`
	EncryptedKeys map[string]string `json:"encrypted_keys"` // userID -> base64 encrypted key
}

// Request/Response payloads for data fetching over WS

type RequestEvent struct {
	ID      string          `json:"id"`
	Action  string          `json:"action"`
	Payload json.RawMessage `json:"payload"`
}

type ResponseEvent struct {
	ID      string      `json:"id"`
	Action  string      `json:"action"`
	OK      bool        `json:"ok"`
	Payload interface{} `json:"payload,omitempty"`
	Error   string      `json:"error,omitempty"`
}

// Request action constants
const (
	ActionSyncInit       = "sync:init"
	ActionMessageList    = "messages:list"
	ActionThreadList     = "threads:list"
	ActionThreadMessages = "threads:messages"
	ActionDMList         = "dms:list"
	ActionDMMessages     = "dms:messages"
)

type MessagesListRequest struct {
	ChannelID string `json:"channel_id"`
	Before    string `json:"before"`
	Limit     int    `json:"limit"`
}

type ThreadsListRequest struct {
	ChannelID string `json:"channel_id"`
}

type ThreadMessagesRequest struct {
	ThreadID string `json:"thread_id"`
	Before   string `json:"before"`
	Limit    int    `json:"limit"`
}

type DMMessagesRequest struct {
	DMID   string `json:"dm_id"`
	Before string `json:"before"`
	Limit  int    `json:"limit"`
}

// Reaction client -> server events
const (
	EventReactionAdd    = "reaction:add"
	EventReactionRemove = "reaction:remove"
)

type ReactionRequestPayload struct {
	ChannelID string `json:"channel_id"`
	MessageID string `json:"message_id"`
	Emoji     string `json:"emoji"`
}

File: server-rs/src/ws/events.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

// ── Client → Server event types ─────────────────────────────────────────────
pub const EVENT_MESSAGE_SEND: &str = "message:send";
pub const EVENT_MESSAGE_EDIT: &str = "message:edit";
pub const EVENT_MESSAGE_DELETE: &str = "message:delete";
pub const EVENT_TYPING_START: &str = "typing:start";
pub const EVENT_TYPING_STOP: &str = "typing:stop";
pub const EVENT_PRESENCE_UPDATE: &str = "presence:update";
pub const EVENT_CHANNEL_JOIN: &str = "channel:join";
pub const EVENT_CHANNEL_LEAVE: &str = "channel:leave";
pub const EVENT_THREAD_MESSAGE_SEND: &str = "thread:message:send";
pub const EVENT_THREAD_MESSAGE_EDIT: &str = "thread:message:edit";
pub const EVENT_THREAD_MESSAGE_REMOVE: &str = "thread:message:remove";
pub const EVENT_VOICE_JOIN: &str = "voice:join";
pub const EVENT_VOICE_LEAVE: &str = "voice:leave";
pub const EVENT_VOICE_ANSWER: &str = "voice:answer";
pub const EVENT_VOICE_ICE_CANDIDATE: &str = "voice:ice-candidate";
pub const EVENT_VOICE_MUTE: &str = "voice:mute";
pub const EVENT_VOICE_DEAFEN: &str = "voice:deafen";
pub const EVENT_VOICE_SCREEN_START: &str = "voice:screen-start";
pub const EVENT_VOICE_SCREEN_STOP: &str = "voice:screen-stop";
pub const EVENT_VOICE_WEBCAM_START: &str = "voice:webcam-start";
pub const EVENT_VOICE_WEBCAM_STOP: &str = "voice:webcam-stop";
pub const EVENT_VOICE_KEY_DISTRIBUTE: &str = "voice:key-distribute";
pub const EVENT_REQUEST: &str = "request";
pub const EVENT_PING: &str = "ping";
pub const EVENT_REACTION_ADD: &str = "reaction:add";
pub const EVENT_REACTION_REMOVE: &str = "reaction:remove";

// DM events
pub const EVENT_DM_MESSAGE_SEND: &str = "dm:message:send";
pub const EVENT_DM_MESSAGE_EDIT: &str = "dm:message:edit";
pub const EVENT_DM_MESSAGE_DELETE: &str = "dm:message:delete";
pub const EVENT_DM_TYPING_START: &str = "dm:typing:start";
pub const EVENT_DM_TYPING_STOP: &str = "dm:typing:stop";

// ── Server → Client event types ─────────────────────────────────────────────
pub const EVENT_MESSAGE_NEW: &str = "message:new";
pub const EVENT_MESSAGE_UPDATED: &str = "message:updated";
pub const EVENT_MESSAGE_DELETED: &str = "message:deleted";
pub const EVENT_TYPING_INDICATOR: &str = "typing:indicator";
pub const EVENT_PRESENCE_CHANGED: &str = "presence:changed";
pub const EVENT_MEMBER_JOINED: &str = "member:joined";
pub const EVENT_MEMBER_LEFT: &str = "member:left";
pub const EVENT_CHANNEL_CREATED: &str = "channel:created";
pub const EVENT_CHANNEL_UPDATED: &str = "channel:updated";
pub const EVENT_CHANNEL_DELETED: &str = "channel:deleted";
pub const EVENT_ERROR: &str = "error";
pub const EVENT_PONG: &str = "pong";

// Thread server events
pub const EVENT_THREAD_CREATED: &str = "thread:created";
pub const EVENT_THREAD_MESSAGE_NEW: &str = "thread:message:new";
pub const EVENT_THREAD_MESSAGE_UPDATED: &str = "thread:message:updated";
pub const EVENT_THREAD_MESSAGE_DELETED: &str = "thread:message:deleted";
pub const EVENT_THREAD_UPDATED: &str = "thread:updated";

// Reaction server events
pub const EVENT_REACTION_ADDED: &str = "reaction:added";
pub const EVENT_REACTION_REMOVED: &str = "reaction:removed";

// Voice server events
pub const EVENT_VOICE_OFFER: &str = "voice:offer";
pub const EVENT_VOICE_ICE_OUT: &str = "voice:ice-candidate";
pub const EVENT_VOICE_USER_JOINED: &str = "voice:user-joined";
pub const EVENT_VOICE_USER_LEFT: &str = "voice:user-left";
pub const EVENT_VOICE_SPEAKING: &str = "voice:speaking";
pub const EVENT_VOICE_STATE: &str = "voice:state";
pub const EVENT_VOICE_MUTE_UPDATE: &str = "voice:mute-update";
pub const EVENT_VOICE_SCREEN_UPDATE: &str = "voice:screen-update";
pub const EVENT_VOICE_WEBCAM_UPDATE: &str = "voice:webcam-update";

// DM server events
pub const EVENT_DM_MESSAGE_NEW: &str = "dm:message:new";
pub const EVENT_DM_MESSAGE_UPDATED: &str = "dm:message:updated";
pub const EVENT_DM_MESSAGE_DELETED: &str = "dm:message:deleted";
pub const EVENT_DM_CREATED: &str = "dm:created";

// ── Request/Response action types ───────────────────────────────────────────
pub const ACTION_SYNC_INIT: &str = "sync:init";
pub const ACTION_MESSAGE_LIST: &str = "messages:list";
pub const ACTION_THREAD_LIST: &str = "threads:list";
pub const ACTION_THREAD_MESSAGES: &str = "threads:messages";
pub const ACTION_DM_LIST: &str = "dms:list";
pub const ACTION_DM_MESSAGES: &str = "dms:messages";

// ── Event struct ────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Event {
    #[serde(rename = "type")]
    pub event_type: String,
    #[serde(default)]
    pub payload: Value,
}

impl Event {
    pub fn new(event_type: &str, payload: impl Serialize) -> Result<Self, serde_json::Error> {
        Ok(Event {
            event_type: event_type.to_string(),
            payload: serde_json::to_value(payload)?,
        })
    }

    pub fn to_bytes(&self) -> Result<Vec<u8>, serde_json::Error> {
        serde_json::to_vec(self)
    }
}

// ── Payload types ───────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MessageSendPayload {
    pub channel_id: String,
    pub content: String,
    #[serde(rename = "type", default)]
    pub msg_type: String,
    #[serde(default)]
    pub thread_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MessageEditPayload {
    pub message_id: String,
    pub channel_id: String,
    pub content: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MessageDeletePayload {
    pub message_id: String,
    pub channel_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MessageNewPayload {
    pub id: String,
    pub channel_id: String,
    pub author_id: String,
    pub username: String,
    pub content: String,
    #[serde(rename = "type")]
    pub msg_type: String,
    #[serde(default)]
    pub thread_id: String,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TypingPayload {
    pub channel_id: String,
    pub user_id: String,
    pub username: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PresenceUpdatePayload {
    pub user_id: String,
    pub status_type: String,
    pub status_text: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChannelJoinPayload {
    pub channel_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VoiceJoinPayload {
    pub channel_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VoiceAnswerPayload {
    pub channel_id: String,
    pub sdp: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VoiceICECandidatePayload {
    pub channel_id: String,
    pub candidate: String,
    #[serde(default)]
    pub sdp_mid: String,
    #[serde(default)]
    pub sdp_mline_index: u16,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VoiceOfferPayload {
    pub channel_id: String,
    pub sdp: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VoiceMutePayload {
    pub channel_id: String,
    pub muted: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VoiceDeafenPayload {
    pub channel_id: String,
    pub deafened: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VoiceUserJoinedPayload {
    pub channel_id: String,
    pub user_id: String,
    pub username: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VoiceUserLeftPayload {
    pub channel_id: String,
    pub user_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VoiceKeyDistributePayload {
    pub channel_id: String,
    pub sender_id: String,
    pub key_id: u32,
    pub encrypted_keys: std::collections::HashMap<String, String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReactionPayload {
    pub message_id: String,
    pub channel_id: String,
    pub emoji: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReactionEventPayload {
    pub message_id: String,
    pub channel_id: String,
    pub user_id: String,
    pub emoji: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DMMessageNewPayload {
    pub id: String,
    pub dm_channel_id: String,
    pub author_id: String,
    pub username: String,
    pub content: String,
    #[serde(rename = "type")]
    pub msg_type: String,
    pub created_at: String,
}

// Thread payloads
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ThreadMessageSendPayload {
    pub thread_id: String,
    pub content: String,
    #[serde(default)]
    pub nonce: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ThreadMessageEditPayload {
    pub thread_id: String,
    pub message_id: String,
    pub content: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ThreadMessageRemovePayload {
    pub thread_id: String,
    pub message_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ThreadMessageNewPayload {
    pub id: String,
    pub thread_id: String,
    pub channel_id: String,
    pub author_id: String,
    pub content: String,
    #[serde(rename = "type")]
    pub msg_type: String,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ThreadUpdatedPayload {
    pub id: String,
    pub title: String,
    pub message_count: i32,
    pub last_message_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ThreadMessageUpdatedPayload {
    pub id: String,
    pub thread_id: String,
    pub content: String,
    pub edited_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ThreadMessageDeletedPayload {
    pub id: String,
    pub thread_id: String,
}

// Voice broadcast payloads
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VoiceScreenPayload {
    pub channel_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VoiceScreenUpdatePayload {
    pub channel_id: String,
    pub user_id: String,
    pub sharing: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VoiceWebcamUpdatePayload {
    pub channel_id: String,
    pub user_id: String,
    pub sharing: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VoiceMuteUpdatePayload {
    pub channel_id: String,
    pub user_id: String,
    pub muted: bool,
    pub deafened: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VoiceStatePayload {
    pub channel_id: String,
    pub peers: Vec<crate::voice::VoicePeer>,
}

// DM payloads
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DMMessageSendPayload {
    pub dm_channel_id: String,
    pub content: String,
    #[serde(rename = "type", default)]
    pub msg_type: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DMMessageEditPayload {
    pub dm_channel_id: String,
    pub message_id: String,
    pub content: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DMMessageDeletePayload {
    pub dm_channel_id: String,
    pub message_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DMTypingPayload {
    pub dm_channel_id: String,
}

// Request/Response
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RequestEvent {
    pub id: String,
    pub action: String,
    #[serde(default)]
    pub payload: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResponseEvent {
    pub id: String,
    pub action: String,
    pub ok: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub payload: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}
